### PR TITLE
feat(indexer): search output redesign — pretty/simple/table formatters, confidence, highlighting

### PR DESCRIPTION
**Pull request by genesiscz**
- **Confidence normalization** — cosine/rrf/bm25 scores mapped to 0-100% percentages
- **Display name cleanup** — `MyClass (part 3)` → `MyClass:10-45` with line ranges
- **Query word highlighter** — ANSI-colorized query terms in search results
- **Three output formatters** — `pretty` (default, syntax-highlighted), `simple` (grep-like), `table` (compact)
- **"semantic" mode alias** — maps to `vector` mode silently
- **Interactive rebuild** — prompts for driver migration, brute-force warning
- **Comprehensive tests** — confidence, display-name, highlight, search-output, search-output-formats
Plan: `.claude/plans/2026-03-25-IndexerSearchOutputRedesign.md`
- [ ] `tools indexer search "function" --format pretty` shows colorized output with confidence %
- [ ] `tools indexer search "function" --format simple` shows grep-like one-line output
- [ ] `tools indexer search "function" --format table` shows compact table
- [ ] `tools indexer search "function" --mode semantic` works as alias for vector
- [ ] `bunx tsgo --noEmit` passes with zero errors

## Summary by CodeRabbit

* **New Features**
  * Enhanced rebuild command with interactive action selection menu for reindex, reembed, and migration options.
  * Added `--mode` option to search command for explicit search type control.
  * Added `--confidence` filtering option to search results.
  * Expanded search output formats: `pretty`, `simple`, `table`, `json`, and `toon`.
  * Search results now display confidence percentages alongside match results.

* **Documentation**
  * Updated vector store warning message to guide users toward performance improvements.

### CODE REVIEW

File: src/indexer/commands/rebuild.ts [hunk 1]
```diff
@@ -46,17 +46,44 @@ export function registerRebuildCommand(program: Command): void {
...
                         return;
                     }
+
+                    if (action === "migrate-driver") {
+                        p.log.warn("Driver migration not yet implemented — running full reindex instead");
+                    }
+
+                    if (action === "reembed") {
+                        p.log.warn("Re-embedding not yet implemented — running full reindex instead");
+                    }
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/indexer/commands/rebuild.ts]: /gemini Fixed in [1195bc2](https://github.com/genesiscz/GenesisTools/commit/1195bc2cb01744cd2cb552a0581584be7f530073) — unimplemented actions now early-return instead of falling through to full rebuild.

File: src/indexer/commands/search.ts [hunk 2]
```diff
@@ -31,93 +27,153 @@ export function registerSearchCommand(program: Command): void {
...
+
+                        allResults.sort((a, b) => b.result.score - a.result.score);
+
+                        if (opts.file) {
+                            allResults = allResults.filter((r) => r.result.doc.filePath.includes(opts.file!));
+                        }
+
+                        allResults = allResults.slice(0, limit);
+                    }
                 }
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/indexer/commands/search.ts]: /gemini Fixed in [1195bc2](https://github.com/genesiscz/GenesisTools/commit/1195bc2cb01744cd2cb552a0581584be7f530073) — extracted searchAndCollect helper to deduplicate the search/sort/filter/slice logic.

File: src/indexer/commands/search.ts [hunk 3]
```diff
@@ -31,93 +27,153 @@ export function registerSearchCommand(program: Command): void {
...
-                    console.log(SafeJSON.stringify(output, null, 2));
-                    return;
-                }
+                        file: r.filePath,
+                        name: r.displayName,
+                        language: r.language,
+                        confidence: r.confidence,
+                        method: r.method,
+                        lines: `${r.startLine}-${r.endLine}`,
+                        preview: r.content.replace(/\n/g, " ").replace(/\s+/g, " ").trim().slice(0, 200),
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/indexer/commands/search.ts]: /gemini Fixed in [1195bc2](https://github.com/genesiscz/GenesisTools/commit/1195bc2cb01744cd2cb552a0581584be7f530073) — now using truncateText utility which appends ellipsis when content is truncated.

File: src/indexer/commands/rebuild.ts [hunk 4]
```diff
@@ -46,17 +46,44 @@ export function registerRebuildCommand(program: Command): void {
 
                 const metas = manager.listIndexes();
                 const meta = metas.find((m) => m.name === targetName);
-                const chunkCount = meta?.stats.totalChunks ?? 0;
 
-                if (process.stdout.isTTY && chunkCount > 0) {
-                    const confirmed = await p.confirm({
-                        message: `Rebuild "${targetName}" (${chunkCount.toLocaleString()} chunks)? This will re-scan all source files.`,
+                if (process.stdout.isTTY && process.stdin.isTTY) {
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/indexer/commands/rebuild.ts]: @coderabbitai Fixed in [1195bc2](https://github.com/genesiscz/GenesisTools/commit/1195bc2cb01744cd2cb552a0581584be7f530073) — replaced raw TTY checks with isInteractive() from utils/cli.

File: src/indexer/commands/rebuild.ts [hunk 5]
```diff
@@ -46,17 +46,44 @@ export function registerRebuildCommand(program: Command): void {
...
+                        { value: "reembed", label: "Re-embed all chunks (regenerate vectors, keep content)" },
+                        { value: "reindex-reembed", label: "Full reindex + re-embed" },
+                    ];
+
+                    if (currentDriver !== "sqlite-vec") {
+                        options.splice(1, 0, {
+                            value: "migrate-driver",
+                            label: `Migrate vector storage: ${currentDriver} → sqlite-vec`,
+                        });
+                    }
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/indexer/commands/rebuild.ts]: @coderabbitai Fixed in [1195bc2](https://github.com/genesiscz/GenesisTools/commit/1195bc2cb01744cd2cb552a0581584be7f530073) — no longer defaults to sqlite-brute when vectorDriver is unset. Only shows migration option when an explicit non-sqlite-vec driver is configured.

File: src/indexer/commands/rebuild.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/indexer/commands/rebuild.ts]: @coderabbitai Fixed in [1195bc2](https://github.com/genesiscz/GenesisTools/commit/1195bc2cb01744cd2cb552a0581584be7f530073) — unimplemented actions now early-return instead of falling through to rebuildIndex.

File: src/indexer/lib/highlight.ts [hunk 6]
```diff
@@ -0,0 +1,40 @@
+import { createColors } from "picocolors";
+
+const pc = createColors(true);
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/indexer/lib/highlight.ts]: @coderabbitai Fixed in [1195bc2](https://github.com/genesiscz/GenesisTools/commit/1195bc2cb01744cd2cb552a0581584be7f530073) — switched to standard picocolors import (auto-detect) in both highlight.ts and search-output.ts. Tests updated to be color-support aware.

File: src/indexer/lib/highlight.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    return result;
+}
+
+export function highlightQueryWords(text: string, words: string[]): string {
+    if (!text || words.length === 0) {
+        return text;
+    }
+
+    const pattern = words.map(escapeRegex).join("|");
+    const regex = new RegExp(`(${pattern})`, "gi");
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/indexer/lib/highlight.ts]: @coderabbitai Fixed in [1195bc2](https://github.com/genesiscz/GenesisTools/commit/1195bc2cb01744cd2cb552a0581584be7f530073) — words are now sorted by length descending before building the regex alternation.

File: src/indexer/lib/search-mode.test.ts [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+    });
+
+    it("maps 'semantic' to 'vector'", () => {
+        expect(resolveSearchMode("semantic")).toBe("vector");
+    });
+
+    it("returns undefined for unknown modes", () => {
+        expect(resolveSearchMode("banana")).toBeUndefined();
+    });
+});
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/indexer/lib/search-mode.test.ts:25]: @coderabbitai Fixed in [1195bc2](https://github.com/genesiscz/GenesisTools/commit/1195bc2cb01744cd2cb552a0581584be7f530073) — added detectMode tests for both embeddingCount === 0 and embeddingCount > 0.

File: src/indexer/commands/search.ts [hunk 9]
```diff
@@ -1,72 +1,24 @@
...
-        }
-    }
-
-    allResults.sort((a, b) => b.result.score - a.result.score);
-    allResults = filterByMinScore(allResults, mode);
-    allResults = allResults.slice(0, limit);
-
-    return allResults;
+    confidence?: number;
+    contextChunks?: number;
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/indexer/commands/search.ts]: @coderabbitai Fixed in [27b8823](https://github.com/genesiscz/GenesisTools/commit/27b8823b68d4b38747647e3bd992c2b98bf9e6ed) — removed --context-chunks flag since it is not wired up yet. Will re-add when the feature is implemented.

File: src/indexer/commands/search.ts [hunk 10]
```diff
@@ -77,65 +29,86 @@ export function registerSearchCommand(program: Command): void {
         .option("-i, --index <name>", "Search specific index (default: all)")
         .option("-m, --mode <mode>", "Search mode: fulltext, vector, hybrid (default: auto-detect)")
         .option("-l, --limit <n>", "Max results", parseInt, 20)
-        .option("-f, --file <filter>", "Filter results to files matching this substring (e.g. '.ts', 'src/utils')")
-        .option("--format <type>", "Output format: table, json, toon (default: table)")
+        .option("-f, --file <filter>", "Filter results to files matching this substring")
+        .option("--format <type>", "Output format: pretty, simple, table, json, toon (default: pretty)")
+        .option("-c, --confidence <min>", "Minimum confidence % (0-100)", parseInt)
+        .option("--context-chunks <n>", "Show N surrounding chunks for context", parseInt)
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/indexer/commands/search.ts]: @coderabbitai Fixed in [27b8823](https://github.com/genesyscz/GenesisTools/commit/27b8823b68d4b38747647e3bd992c2b98bf9e6ed) — unknown --format values are now rejected early with a clear error message listing valid options.

File: src/indexer/commands/search.ts [hunk 11]
```diff
@@ -77,65 +29,86 @@ export function registerSearchCommand(program: Command): void {
...
+                        allResults.sort((a, b) => b.result.score - a.result.score);
 
-                        if (allResults.length > 0) {
-                            p.log.info(pc.dim("No keyword matches — showing semantic results instead"));
+                        if (opts.file) {
+                            allResults = allResults.filter((r) => r.result.doc.filePath.includes(opts.file!));
                         }
+
+                        allResults = allResults.slice(0, limit);
                     }
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/indexer/commands/search.ts]: @coderabbitai Fixed in [27b8823](https://github.com/genesiscz/GenesisTools/commit/27b8823b68d4b38747647e3bd992c2b98bf9e6ed) — hybrid fallback now only triggers when mode was auto-detected, not when user explicitly passed --mode fulltext.

File: src/indexer/lib/confidence.test.ts [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+            expect(mid).toBeGreaterThanOrEqual(0);
+            expect(mid).toBeLessThanOrEqual(100);
+
+            expect(high).toBeGreaterThanOrEqual(0);
+            expect(high).toBeLessThanOrEqual(100);
+
+            expect(low).toBeLessThan(mid);
+            expect(mid).toBeLessThan(high);
+        });
+    });
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/indexer/lib/confidence.test.ts:81]: @coderabbitai Fixed in [27b8823](https://github.com/genesiscz/GenesisTools/commit/27b8823b68d4b38747647e3bd992c2b98bf9e6ed) — added explicit test for maxScore===0 edge case verifying it falls back to the heuristic and preserves ordering.

File: src/indexer/lib/display-name.ts [hunk 13]
```diff
@@ -0,0 +1,29 @@
+export function stripPartSuffixes(name: string): string {
+    return name.replace(/\s*\(part\s+\d+\)/g, "").trim();
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/indexer/lib/display-name.ts]: @coderabbitai Fixed in [27b8823](https://github.com/genesiscz/GenesisTools/commit/27b8823b68d4b38747647e3bd992c2b98bf9e6ed) — regex now anchored to end of string, so internal occurrences like RFC (part 2) parser are preserved.

File: src/indexer/lib/search-output.ts [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+                parts.push(`${pc.dim(`${lineNum}|`)}${contentLines[i]}`);
+            } else {
+                parts.push(`${pc.dim("|")}${contentLines[i]}`);
+            }
+        }
+
+        parts.push("");
+    }
+
+    return parts.join("\n").trimEnd();
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/indexer/lib/search-output.ts:131]: @coderabbitai By design — simple format is intentionally the detailed line-numbered output for grep-like inspection. table format serves as the compact single-line-per-result view. Users control volume via --limit. The three formats serve different use cases: pretty (rich TTY), simple (detailed with line numbers), table (compact overview).

File: src/indexer/commands/rebuild.ts [hunk 15]
```diff
@@ -46,17 +47,46 @@ export function registerRebuildCommand(program: Command): void {
...
+                        { value: "reembed", label: "Re-embed all chunks (regenerate vectors, keep content)" },
+                        { value: "reindex-reembed", label: "Full reindex + re-embed" },
+                    ];
+
+                    if (currentDriver && currentDriver !== "sqlite-vec") {
+                        options.splice(1, 0, {
+                            value: "migrate-driver",
+                            label: `Migrate vector storage: ${currentDriver} → sqlite-vec`,
+                        });
+                    }
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/indexer/commands/rebuild.ts]: @coderabbitai Fixed in [6b12dc7](https://github.com/genesiscz/GenesisTools/commit/6b12dc7c122587d2bab77acfee1ec3c594cae459) — migration option now only shown for sqlite-brute, not qdrant or other external stores.

File: src/indexer/commands/search.ts [hunk 16]
```diff
@@ -77,65 +60,53 @@ export function registerSearchCommand(program: Command): void {
         .option("-i, --index <name>", "Search specific index (default: all)")
         .option("-m, --mode <mode>", "Search mode: fulltext, vector, hybrid (default: auto-detect)")
         .option("-l, --limit <n>", "Max results", parseInt, 20)
-        .option("-f, --file <filter>", "Filter results to files matching this substring (e.g. '.ts', 'src/utils')")
-        .option("--format <type>", "Output format: table, json, toon (default: table)")
+        .option("-f, --file <filter>", "Filter results to files matching this substring")
+        .option("--format <type>", "Output format: pretty, simple, table, json, toon (default: pretty)")
+        .option("-c, --confidence <min>", "Minimum confidence % (0-100)", parseInt)
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/indexer/commands/search.ts:64]: @coderabbitai Fixed in [6b12dc7](https://github.com/genesiscz/GenesisTools/commit/6b12dc7c122587d2bab77acfee1ec3c594cae459) — --confidence now validated early for NaN, negatives, and values exceeding 100 with a clear error message.

File: src/indexer/commands/search.ts [hunk 17]
```diff
@@ -144,50 +115,63 @@ export function registerSearchCommand(program: Command): void {
...
-                    return;
-                }
+                        file: r.filePath,
+                        name: r.displayName,
+                        language: r.language,
+                        confidence: r.confidence,
+                        method: r.method,
+                        lines: `${r.startLine}-${r.endLine}`,
+                        preview: truncateText(r.content.replace(/\n/g, " ").replace(/\s+/g, " ").trim(), 200),
+                    }));
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/indexer/commands/search.ts:163]: @coderabbitai Fixed in [6b12dc7](https://github.com/genesiscz/GenesisTools/commit/6b12dc7c122587d2bab77acfee1ec3c594cae459) — lines field now outputs null instead of undefined-undefined when line numbers are missing.

File: src/indexer/lib/search-output.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/indexer/lib/search-output.ts:131]: @coderabbitai By design — simple format is the detailed line-numbered output for grep-like inspection. table is the compact format. Users control volume via --limit.

File: src/indexer/commands/rebuild.ts [hunk 18]
```diff
@@ -46,17 +47,46 @@ export function registerRebuildCommand(program: Command): void {
...
+                        { value: "reembed", label: "Re-embed all chunks (regenerate vectors, keep content)" },
+                        { value: "reindex-reembed", label: "Full reindex + re-embed" },
+                    ];
+
+                    if (currentDriver === "sqlite-brute") {
+                        options.splice(1, 0, {
+                            value: "migrate-driver",
+                            label: `Migrate vector storage: ${currentDriver} → sqlite-vec`,
+                        });
+                    }
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/indexer/commands/rebuild.ts]: @coderabbitai Fixed in [0845e3b](https://github.com/genesiscz/GenesisTools/commit/0845e3b95956412bd7e6451c65bfe669d25c7495) — removed unimplemented options (migrate-driver, reembed, reindex-reembed) from the selector. Simplified to a single confirm prompt for rebuild. Will re-add action options when the manager layer supports them.

File: src/indexer/lib/display-name.test.ts [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+    });
+
+    it("falls back to just lines when no name or kind", () => {
+        expect(formatChunkDisplayName(undefined, 1, 30)).toBe("L1-30");
+    });
+
+    it("handles single-line chunk", () => {
+        expect(formatChunkDisplayName("handler", 42, 42)).toBe("handler:42");
+    });
+});
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/indexer/lib/display-name.test.ts:55]: @coderabbitai Fixed in [0845e3b](https://github.com/genesiscz/GenesisTools/commit/0845e3b95956412bd7e6451c65bfe669d25c7495) — added tests for undefined and NaN line inputs verifying fallback to cleaned name, kind, or chunk.

File: src/indexer/lib/highlight.test.ts [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+    });
+
+    it("filters short words (<=2 chars)", () => {
+        const result = parseQueryWords("how is the AI working");
+        expect(result).toContain("how");
+        expect(result).toContain("the");
+        expect(result).toContain("working");
+        expect(result).not.toContain("is");
+        expect(result).not.toContain("AI");
+    });
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/indexer/lib/highlight.test.ts:25]: @coderabbitai Fixed in [0845e3b](https://github.com/genesiscz/GenesisTools/commit/0845e3b95956412bd7e6451c65bfe669d25c7495) — assertion now checks for lowercase ai to match parseQueryWords output.

File: src/indexer/lib/search-output-formats.test.ts [hunk 21]
```diff
@@ -0,0 +1,405 @@
...
+                results: [makeResult()],
+                format: "simple",
+                ...defaultOpts,
+                highlightWords: ["reservation"],
+            });
+            const plain = stripAnsi(output);
+
+            expect(output.length).toBeGreaterThan(plain.length);
+            expect(plain).toContain("reservation");
+        });
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/indexer/lib/search-output-formats.test.ts:176]: @coderabbitai Fixed in [0845e3b](https://github.com/genesiscz/GenesisTools/commit/0845e3b95956412bd7e6451c65bfe669d25c7495) — highlight tests now compare output with and without highlightWords to verify the highlighting itself adds ANSI codes, not just other styling.
